crates/resolver/src/lookup_ip.rs

```diff
@@ -317,6 +330,74 @@ impl<C: DnsHandle> LookupContext<C> {
     }
 }
 
+pin_project! {
+    struct RacingLookup<A, Aaaa, F> {
+        #[pin]
+        a: A,
+        a_output: Option<Result<Lookup, ProtoError>>,
+        #[pin]
+        aaaa: Aaaa,
+        aaaa_output: Option<Result<Lookup, ProtoError>>,
+        delay: Option<F>,
+        sleeping: Option<Pin<Box<dyn Future<Output = ()> + Send>>>,
+    }
+}
+
+impl<A, Aaaa, F> Future for RacingLookup<A, Aaaa, F>
+where
+    A: Future<Output = Result<Lookup, ProtoError>>,
+    Aaaa: Future<Output = Result<Lookup, ProtoError>>,
+    F: FnOnce() -> Pin<Box<dyn Future<Output = ()> + Send>> + Send,
+{
+    type Output = Result<Lookup, ProtoError>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        // Kick off the AAAA query first
+        if this.aaaa_output.is_none() {
+            match this.aaaa.poll(cx) {
+                Poll::Ready(Ok(aaaa)) if !aaaa.is_empty() => return Poll::Ready(Ok(aaaa)),
```

Comment:
if the AAAA is empty, could this cause a deadlock in the case where there is no delay and there is also no A?